tests/0101-fetch-from-follower.cpp

```diff
@@ -0,0 +1,294 @@
+  delete tconf; // TODO: correct?
+  test_assert(topic, tostr() << "Failed to create topic: " << errstr);
+
+  err = p->metadata(0, topic, &metadata, 5000);  // TODO: best timeout?
+  test_assert(err == RdKafka::ERR_NO_ERROR,
+    tostr() <<  "%% Failed to acquire metadata: "
+            << RdKafka::err2str(err));
+
+  test_assert(metadata->topics()->size() == 1,
+    tostr() << "expecting metadata for exactly one topic. "
+            << "have metadata for " << metadata->topics()->size()
+            << "topics");
+
+  RdKafka::Metadata::TopicMetadataIterator topicMetadata = metadata->topics()->begin();
+  RdKafka::TopicMetadata::PartitionMetadataIterator partitionMetadata = (*topicMetadata)->partitions()->begin();
+
+  *leader = (*partitionMetadata)->leader();
+
+  size_t idx = 0;
+  RdKafka::PartitionMetadata::ReplicasIterator replicasIterator;
+  for (replicasIterator = (*partitionMetadata)->replicas()->begin();
+           replicasIterator != (*partitionMetadata)->replicas()->end();
+           ++replicasIterator) {
+    brokers.push_back(*replicasIterator);
+    idx++;
+  }
+
+  delete metadata;
+  delete topic;
+  delete p;
+}
+
+
+static void do_fff_test (void) {
+  size_t i;
+  int cnt;
+  int msgcnt = 1000;
+  const int msgsize = 100;
+  std::vector<int> replica_ids;
+  int leader_id;
+  int client_rack_id;
+  std::string client_rack, topic_str;
+  RdKafka::Conf *conf;
+  RdKafka::KafkaConsumer *c;
+  std::vector<std::string> topics;
+  std::string errstr;
+  RdKafka::ErrorCode err;
+
+  /* Produce some messages to a single partition topic
+   * with 3 replicas.
+   */
+  topic_str = Test::mk_topic_name("0101-fetch-from-follower", 1);
+  test_create_topic(NULL, topic_str.c_str(), 1, 3);
+  test_produce_msgs_easy_size(topic_str.c_str(), 0, 0, msgcnt, msgsize);
+
+  get_brokers_info(topic_str, &leader_id, replica_ids);
+  test_assert(replica_ids.size() == 3, tostr() << "expecting three replicas, but " << replica_ids.size() << " were reported.");
+  Test::Say(tostr() << topic_str << " leader id: " << leader_id << ", all replica ids: [" << replica_ids[0] << ", " << replica_ids[1] << ", " << replica_ids[2] << "]\n");
+
+  /* arrange for the consumer's client.rack to align with a broker that is not the leader. */
+  for (i=0; i<replica_ids.size(); ++i) {
+    if (replica_ids[i] != leader_id) {
+      client_rack_id = replica_ids[i];
+      break;
+    }
+  }
+  client_rack = tostr() << "RACK" << client_rack_id;
+  Test::Say(tostr() << "client.rack: " << client_rack << "\n");
+
+  Test::conf_init(&conf, NULL, 10);
+  Test::conf_set(conf, "group.id", topic_str);
+  Test::conf_set(conf, "auto.offset.reset", "earliest");
+  Test::conf_set(conf, "enable.auto.commit", "false");
+  Test::conf_set(conf, "statistics.interval.ms", "1000");
+  conf->set("event_cb", &ex_event_cb, errstr);
+  Test::conf_set(conf, "client.rack", client_rack);
+  Test::conf_set(conf, "debug", "all");
```

Comment:
Remove. Use `TEST_DEBUG=all TESTS=0101 make` instead.

Reply:
i need access to a good reference on how to drive your test env.

Reply:
Not addressed

Reply:
what happened here is that i addressed it, then unaddressed it because that was easier than remembering the env var.